lntest/itest/lnd_amp_test.go

```diff
@@ -245,6 +254,24 @@ func testSendToRouteAMP(net *lntest.NetworkHarness, t *harnessTest) {
 		}()
 	}
 
+	// Send the first shard, this cause Bob to JIT add an invoice.
+	sendShard(0, sendRoutes[0])
+
+	// Ensure we get a notification of the invoice being added by Bob.
```

Comment:
Given we get this "for free" essentially right now, what do you think of reviving my old idea to give users/programs an option to create an AMP invoice that doesn't have a payment address set? The docs would instruct users to set this value if they're going after the donation use case and don't necessarily need to associate the payment with a given identifier (like an accountID). In order to allow the receipeint to distinguish these payments from a cetegorical persepective, we could thread through the description or description hash from the invoice to them using a TLV record. We store the set of TLV records associated w/ a payment in the HTLC so this can easily be extracted by the receiver for record keeping purposes. 

Thoughts @halseth? 

This would give us a simple variant of re-useable invoice w/o going through all the extra DB work to add an extra contract state per invoice. IMO if we don't add something simple like this at this point, and ship with it, then a lot of users will be disappointed as being able to re-use an invoice is the main selling point for them. Otherwise it feels like somewhat of a lost opportunity as users will go to design applications/services that assume invoice re-use, but then will find that functionally from their PoV, it's no different than the existing regular invoice format.

Reply:
I would be cautious to do another keysend-style shortcut, because it may hold back people from moving to recurring AMP once it is ready. After all they've got their reusable donation invoices working already then.

Even with keysend today, you could do discernible donations by adding a tlv record. But I suspect it isn't happening because there is no standard for it. If you build this AMP-TLV-identifier into `lnd`, it is much easier to build on it. And much harder to get rid of it again eventually.

My view on it is to just press on and get recurring done for an early 0.14 release :rocket:

Reply:
> because it may hold back people from moving to recurring AMP once it is ready. After all they've got their reusable donation invoices working already then.

The idea is that from the user's PoV things (when sending) things wouldn't need to change much. In the case of keysend, they actually need to _modify_ their sending behavior, but in this case they'd just hit "send" with the same invoice. What _would_ change though is correlation at the receiver side (which of these payments are for my Github donation profile) though. 

IMO we'd be setting up a lot of users for disappointment if they go to re-use the invoice and get an error that it isn't possible. Yeah this is a bit of a hack/stop-gap, but IMO it's worth it as we haven't even really started to fully explore what the "ideal" receiver side API for tracking these things looks like, not to mention the possible changes in the invoice/HTLC storage to make retrieval/scanning of the information efficient.

Reply:
>The idea is that from the user's PoV things (when sending) things wouldn't need to change much. 

Yes, the send API doesn't change, but I am more concerned about what happens on the wire with an additional TLV field to identify the payment. That is the part the probably needs to be supported for quite a while in applications. For as long as donations sent from old clients keep coming in.

>MO we'd be setting up a lot of users for disappointment if they go to re-use the invoice and get an error that it isn't possible.

I think this mainly has to do with communication. There are many ways (docs, lncli, release notes, etc) to make clear that AMP in its current form is mostly multi-path keysend.

>explore what the "ideal" receiver side API for tracking these things looks like

And how about keeping the flat model, but just making (paymentaddr, setid) the index? The API will be mostly the same and perhaps not ideal because it doesn't really contain that nested structure, but is that any different from how users would query invoices with the extra-tlv-shortcut that you describe above? At least AMP is used as intended on the wire level.

Reply:
Yeah I see what you're saying, 0.14 as it's shaping out _should_ be much smaller than past releases, so we should be able to really focus and hone in on the design space. FWIW in the end, what I've described can be implemented as a command line mode (the invoice is basically just to transmit the destination) and can also be done over the RPC system. 

> but is that any different from how users would query invoices with the extra-tlv-shortcut that you describe above?

Doesn't sound too different, as with what I'm proposing they'd essentially need to scan _all_ invoices as there'd be no external invoice. The main gap here to the ideal state is to store an invoice state for each known set id, so it's possible that an invoice is settled multiple times.

Reply:
I am wondering how far you could get by just using `payment_addr | set_id` instead of `payment_addr` as the key in the `pay-addr-index` bucket.